### ISSUE TITLE
UnicodeEncodeError when calling puts function with unicode u'\xff' and the stream.encoding is None

### ISSUE THREAD

**Issue by zhaoguixu** [fabric/fabric#1303]
Hi,
When replace the variable <b>s</b> to another value <b>u'\xff'</b>, this test would fail.

<b>[test-case]</b>
```python
@mock_streams('stdout')
def test_puts_with_encoding_type_none_output():
    """
    puts() should print unicode output without a stream encoding
    """
    s = u"string!"
    output.user = True
    sys.stdout.encoding = None
    puts(s, show_prefix=False)
    eq_(sys.stdout.getvalue(), s + "\n")
```
<b>[traceback]</b>
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-4: ordinal not in range(128)

<b>[comments]</b>
The buggy point locates in the function '_encode' in utils.py, you should not simply cast the msg into str by using str(msg).
```python
def _encode(msg, stream):
    if isinstance(msg, unicode) and hasattr(stream, 'encoding') and not stream.encoding is None:
        return msg.encode(stream.encoding)
    else:
        return str(msg)         #<========== buggy point
```

**Comment by zhaoguixu** [comment]
Anyone cares?

**Comment by ppkt** [comment]
Yup, it seems that changing `return str(msg)` to `return msg` solves the problem without breaking tests. But why this cast was necessary at the first place?

**Comment by kaaveland** [comment]
I'm not absolutely sure it makes sense to change this now. It's not going to absolutely break anyone's code, it basically lets you `puts` things which aren't strings and get something sensible out:

```python
>>> class Foo(object):
...   def __str__(self): return "The best Foo in the world"
...
>>> from fabric.api import puts
>>> puts(Foo())
The best Foo in the world
```
I'm guessing there's code out there that uses this (hopeless to google it right now).

I'm not sure I think it's right that not having an encoding set and using unicode will cause `puts` to crash us, though.

**Comment by bitprophet** [comment]
Suspect the least-worst way to address this for now is to simply update the offending line to be:

``` python
try:
    return str(msg) # Original behavior added for 'reasons'
except UnicodeDecodeError:
    return msg # Best-effort fallback
```

This way, whatever antediluvian reasons we had for casting-to-string will continue to function as-is, but the buggy case under discussion would fall back to "well, I dunno what this is, but I can't `str()` it, so I'll just shove it down the pipe as-is".

**Comment by dclong** [comment]
I encountered a similar issue while running the `sudo` method. Is there any progress on this?
